pkg/agent/proxy/proxier_test.go

```diff
@@ -116,110 +299,92 @@ func withoutProxyLoadBalancerIPs(o *proxyOptions) {
 	o.proxyLoadBalancerIPs = false
 }
 
-func NewFakeProxier(routeClient route.Interface, ofClient openflow.Client, nodePortAddresses []net.IP, groupIDAllocator openflow.GroupAllocator, isIPv6 bool, options ...proxyOptionsFn) *proxier {
-	hostname := "localhost"
-	eventBroadcaster := record.NewBroadcaster()
-	recorder := eventBroadcaster.NewRecorder(
-		runtime.NewScheme(),
-		corev1.EventSource{Component: componentName, Host: hostname},
-	)
-
-	ipFamily := corev1.IPv4Protocol
-	if isIPv6 {
-		ipFamily = corev1.IPv6Protocol
-	}
+func withEndpointSlice(o *proxyOptions) {
+	o.endpointSliceEnabled = true
+}
 
+func NewFakeProxier(routeClient route.Interface, ofClient openflow.Client, nodePortAddresses []net.IP, groupIDAllocator openflow.GroupAllocator, isIPv6 bool, options ...proxyOptionsFn) *proxier {
 	o := &proxyOptions{
 		proxyAllEnabled:      false,
 		proxyLoadBalancerIPs: true,
+		endpointSliceEnabled: false,
 	}
 
 	for _, fn := range options {
 		fn(o)
 	}
 
-	p := &proxier{
-		endpointsChanges:         newEndpointsChangesTracker(hostname, false, isIPv6),
-		serviceChanges:           newServiceChangesTracker(recorder, ipFamily, []string{"kube-system/kube-dns", "192.168.1.2"}),
-		serviceMap:               k8sproxy.ServiceMap{},
-		serviceInstalledMap:      k8sproxy.ServiceMap{},
-		endpointsInstalledMap:    types.EndpointsMap{},
-		endpointReferenceCounter: map[string]int{},
-		endpointsMap:             types.EndpointsMap{},
-		groupCounter:             types.NewGroupCounter(groupIDAllocator, make(chan string, 100)),
-		ofClient:                 ofClient,
-		routeClient:              routeClient,
-		serviceStringMap:         map[string]k8sproxy.ServicePortName{},
-		isIPv6:                   isIPv6,
-		nodePortAddresses:        nodePortAddresses,
-		proxyAll:                 o.proxyAllEnabled,
-		proxyLoadBalancerIPs:     o.proxyLoadBalancerIPs,
-		numLocalEndpoints:        map[apimachinerytypes.NamespacedName]int{},
-	}
+	p := NewProxier("localhost",
+		informers.NewSharedInformerFactory(fake.NewSimpleClientset(), 0),
+		ofClient,
+		isIPv6,
+		routeClient,
+		nodePortAddresses,
+		o.proxyAllEnabled,
+		[]string{"kube-system/kube-dns", "192.168.1.2"},
+		o.proxyLoadBalancerIPs,
+		types.NewGroupCounter(groupIDAllocator, make(chan string, 100)))
 	p.runner = k8sproxy.NewBoundedFrequencyRunner(componentName, p.syncProxyRules, time.Second, 30*time.Second, 2)
+	if o.endpointSliceEnabled {
+		p.endpointsChanges = newEndpointsChangesTracker(hostname, o.endpointSliceEnabled, isIPv6)
+	}
 	return p
 }
 
-func testClusterIP(t *testing.T, svcIP net.IP, ep1IP, ep2IP net.IP, isIPv6, nodeLocalInternal bool, extraSvcs []*corev1.Service, extraEps []*corev1.Endpoints) {
+func testClusterIPAdd(t *testing.T,
```

Comment:
I did a quick check, looks like Node related events are not covered, can it be added? or the effort is time-consuming?

Reply:
Yes,  they are no covered. Like `OnEndpointSliceAdd`, `OnServiceAdd`, they are also not covered. Could you give some suggestions to cover them all?

Reply:
Like this functions, I can add ut for this functions, but I need to more time to write them.

Reply:
Sure, I see the overall package coverage is over 70%, so I feel it's OK for now if it requires too much efforts. We can estimate and check in the future.